backend/commons/src/main/java/org/sonarsource/sonarlint/core/commons/util/git/NativeGit.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * SonarLint Core - Commons
+ * Copyright (C) 2016-2025 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.commons.util.git;
+
+import java.net.URI;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.time.Period;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+import org.sonarsource.sonarlint.core.commons.MultiFileBlameResult;
+import org.sonarsource.sonarlint.core.commons.Version;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+import org.sonarsource.sonarlint.core.commons.util.FileUtils;
+
+import static java.lang.String.format;
+
+public class NativeGit {
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+  private static final String GIT_VERSION_OUTPUT_PREFIX = "git version";
+  private static final Period blameHistoryWindow = Period.ofDays(365);
+  private final String executable;
+
+  public NativeGit(String executable) {
+    this.executable = executable;
+  }
+
+  public Optional<Version> version() {
+    var lines = new ArrayList<String>();
+    var success = executeGitCommand(null, lines::add, executable, "--version");
+    return success ? parseGitVersionOutput(lines) : Optional.empty();
+  }
+
+  static Optional<Version> parseGitVersionOutput(List<String> lines) {
+    if (!lines.isEmpty()) {
+      var versionLine = lines.get(0).trim();
+      if (versionLine.startsWith(GIT_VERSION_OUTPUT_PREFIX)) {
+        var actualVersion = versionLine.substring(GIT_VERSION_OUTPUT_PREFIX.length()).trim();
+        // not all version components are numbers, e.g. on Windows '2.49.0.windows.1'
+        var version = actualVersion.split("\\.", 3);
+        if (version.length > 1) {
+          try {
+            // keep only MAJOR and MINOR numbers, it's sufficient for checking support
+            return Optional.of(Version.create(version[0] + "." + version[1]));
+          } catch (Exception e) {
+            // error will be logged below
+          }
+        }
+      }
+    }
```

Comment:
I would suggest doing it with stream/optional API to reduce nesting and checking, creating a sequential path to the result. `findFirst` provides exactly what we need and we keep all of `Optional.of()` and stuff like that under the hood, creating less noise.

```suggestion
    return lines.stream().findFirst()
      .map(String::trim)
      .filter(line -> line.startsWith(GIT_VERSION_OUTPUT_PREFIX))
      .map(line -> Strings.CI.removeStart(line, GIT_VERSION_OUTPUT_PREFIX))
      .map(String::trim)
      .map(actualVersion -> actualVersion.split("\\.", 3))
      .filter(versionParts -> versionParts.length > 1)
      .map(NativeGit::createVersion); // try/catch part to a different method to avoid noise here
```

Reply:
I tried your approach, the end result is maybe even more verbose than before but I like the style, so I will keep it. Another good idea!